capi-adoption/README.md

```diff
@@ -0,0 +1,159 @@
+
+## Proposal for alternative approach to provide Giant Swarm platform with upstream Cluster API
+
+To pick up the key points from current problems and to fully align with upstream in order to simplify our daily work and allow customers more flexibility while improving stability, here are proposed steps forward.
+
+### Let go of releases and multiple versions in same management cluster
+
+Currently the major source of friction with upstream and the source of complexity on our platform is coming from immutable releases and the need to run multiple versions of everything in the same Kubernetes cluster.
+
+Since the upstream Cluster API drastically changes the way how things are configured, we should leverage it, not work against it. Even though we have managed to contribute code upstream to support our use case, the community sees these contributions as workarounds to avoid blocking us. But we’d be always going in a different direction, increasing the effort and cost to fully implement CAPI and be aligned with the project.
+There are less risks for failure when different parts are managed separately and when active change of resources is always a direct consequence of active change on CR[s]. Management cluster operators can be upgraded when there are fixes, but they must not automatically change any of the reconciled resources. When there is a hot security fix for Kubernetes, it’s enough to update only the Kubernetes version. When workload cluster applications need updating, customers can perform it at their will via our App management infrastructure.
+
+Instead of strictly trying to keep control of release components, we should have sane defaults for CR templating. This would be our way to indicate “we have tested these components in this combination and found them to be rock solid, use them”.
+Focusing on templating with sane defaults rather than enforcing certain values would also mean getting rid of admission controllers that we currently use to validate and default CR content, simplifying the Kubernetes configuration.
+
+Allowing customers to customize those sane defaults would ease the testing of new combinations. For example, since upgrading Kubernetes to a new version would be only a matter of changing the required CRs, customers can test their workloads running on the new Kubernetes version on the cluster they decide at their own pace. And they would be testing that change in isolation. There would be no changes in the infrastructure resources or the installed apps.
+
+### Change our mindset when implementing operators
+
+In order to comply with upstream operator behavior, we must change our mindset when implementing operators. Instead of writing logic that proactively changes reconciled resources immediately when the code is executed, we should adopt the approach where only change in the CRs can cause an action on the target resource.
```

Comment:
Can you give an example of this? I thought we preferred level-based reconciliation rather than edge-based. If a controller is down for some reason and misses the change event, how could it eventually end up in the correct state?

Reply:
This is not so much about level-based vs. edge-based reconciliation but rather about general approach on how to determine change of target resource and how to deal with it when detected.

Examples from CAPZ:
- Several infrastructure resources are expected to be ok if they exist with certain name. See: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/master/azure/services/loadbalancers/loadbalancers.go#L85-L137
- If a worker node is deleted, it requires specific `MachineHealthCheck` configuration in order to be replaced automatically.

> how could it eventually end up in the correct state?

It doesn't, within patch updates. But I also don't expect changes within patch updates. In k8s world so far the minor version bump has been considered a major upgrade and this is the case with Cluster API as well. Between those bumps there could be changes to infrastructure as well, but it's fine because the whole upgrade is bigger than just bugfix patches. One can also create infrastructure migrations to be performed at this point, when necessary.

Reply:
I don't think we are changing anything related to level-based vs edge-based reconciliation?